Create tests for sqlite3

perlpunk opened the pull request:
Creating regression tests for sqlite3.
The tests are using the sqlite3 commandline tool in batch mode and interactively.

- Related ticket: https://progress.opensuse.org/issues/53678
- Needles: none
- Verification run: http://10.161.229.247/tests/81

edit: updated link to test

sqozz commented:
Is there a reason why you name some scripts with `.sh` and others with `.bash`?

perlpunk commented:
> Is there a reason why you name some scripts with `.sh` and others with `.bash`?

Good question. I didn't have a reason. The only difference between the scripts is that `run-tests.sh` is an actual script to run, and the others are just sourced, but that's not a reason for the different suffixes.
I'll change that.

foursixnine commented on tests/console/sqlite3.pm:
```diff
@@ -0,0 +1,51 @@
+
+
+sub run {
+    select_console 'root-console';
```
Since you're not peforming any screen based testing, I guess this test could benefit from using  `$self->select_serial_terminal` which is probably a bit faster.

perlpunk replied:
Ah, I didn't know that. Will have a look!

perlpunk replied:
working, thanks!

foursixnine commented on data/sqlite3/init-tables.sql:
```diff
@@ -0,0 +1,23 @@
+CREATE UNIQUE INDEX movie_dirx on director_movie(did,mid);
+
+INSERT INTO directors (name) VALUES ("Jim Jarmusch"), ( "Tim Burton");
+INSERT INTO directors VALUES(3,'Lana Wachowski');
```
Maybe don't mix single and double quotes? :)

perlpunk replied:
OTOH this can count as testing different quoting styles ;-)

foursixnine replied:
Well played... but then could we also test a movie by `Alejandro González Iñárritu`? :D

perlpunk replied:
Added ;-)

perlpunk commented:
> Aside from moving the TAP implementation (just an opinion, not really a must) to a more generic location [...]

I'd love to do that.
There's also this implementation: https://github.com/ingydotnet/test-more-bash
But it requires some other bash libraries (they are included though), and I think it doesn't have a function to compare arrays yet.

I could try to use the above library and add an array comparison, or I could make my implementation more generic.

foursixnine commented on tests/console/sqlite3.pm:
```diff
@@ -0,0 +1,51 @@
+    );
+
+    # Test various commands
+    assert_script_run("prove -v sqlite3/run-tests.bash");
```
Btw openQA has support to parse and display TAP too ^^, so you can save and upload via parse_extra_log( TAP => $file) iir.

perlpunk replied:
Ah, cool! Looking...

perlpunk replied:
working, thanks!

foursixnine commented:
> > Aside from moving the TAP implementation (just an opinion, not really a must) to a more generic location [...]
> 
> I'd love to do that.
> There's also this implementation: https://github.com/ingydotnet/test-more-bash

I didn't know about this one! 

> But it requires some other bash libraries (they are included though), and I think it doesn't have a function to compare arrays yet.
> 
> I could try to use the above library and add an array comparison, or I could make my implementation more generic.

We can do baby steps too :) but as a starter yours looks good...

perlpunk commented:
I created https://github.com/os-autoinst/openQA/pull/2178 to allow all kinds of filename suffixes for TAP source files

perlpunk commented:
I created a PR to add `cmp-array` to @ingydotnet's test-more-bash: https://github.com/ingydotnet/test-more-bash/pull/5

perlpunk commented:
@foursixnine I'm actually thinking of rewriting the run-tests.bash in perl.
I thought bash would be a good idea, because calling commands in bash is nice.
But thinking about it, all I need here is some simple commandline calls and perl's `qx{...}`, and then comparing the exit code and output can be done easier with perl's Test::More. Since we have perl anyway.

But maybe doing this in bash can be useful for other kinds of tests.

foursixnine commented:
> @foursixnine I'm actually thinking of rewriting the run-tests.bash in perl.
> I thought bash would be a good idea, because calling commands in bash is nice.
> But thinking about it, all I need here is some simple commandline calls and perl's `qx{...}`, and then comparing the exit code and output can be done easier with perl's Test::More. Since we have perl anyway.
> 
> But maybe doing this in bash can be useful for other kinds of tests.

It's not a bad idea either. Having the bash interface could motivate more people to try it :) instead of using assert_script_run for every single test/command they wanna run...

perlpunk commented:
@foursixnine Ok, the perl-rewrite was easy.
If using bash, I think it would be a good idea to use the existing test-more-bash instead of writing another port, but I'm waiting for a [bug with diff](https://github.com/ingydotnet/test-more-bash/pull/7) to be fixed, and Ingy and I haven't agreed on how the `cmp-array` function should look like yet.
So to finish my test, I think I should go with the perl version for now.
I'll keep the bash stuff around.

foursixnine commented:
@perlpunk sounds good :)

perlpunk commented:
Ok, I ran perltidy, rebased to master and squashed all commits.

(The bash code is in [branch sqlite3-bash-test-simple](https://github.com/perlpunk/os-autoinst-distri-opensuse/tree/sqlite3-bash-test-simple) in my fork.)

Martchus commented on tests/console/sqlite3.pm:
```diff
@@ -0,0 +1,61 @@
+use warnings;
+use testapi;
+use utils 'zypper_call';
+use Data::Dumper;
```
I guess that is a leftover from debugging.

perlpunk replied:
yep, I always forget to take that out and since it's core, it's usually not a problem.
should I remove it?

Martchus replied:
Yes, remove it. I often enough forget it too :-)

perlpunk commented:
I have to update my test.
I just ran it on sle-12-SP1 where we have an older sqlite version (3.8.10.2), and apparently in this version FOREIGN KEYS pointing to a table won't get updated if you rename that table.
So I have to recreate the keys with the new table name.

perlpunk commented:
Tests are fixed now also for older sqlite3 versions and pass on SLE12 SP1-4 and SLE15.
I added details to the Summary and added the test to `load_extra_tests_console`

okurz commented:
@perlpunk great to see your contribution to tests: I hope I am not destroying your motivation but your test looks like a perfect example that should *not* be in openQA but rather in a testsuite that is called by the %check target in the spec file called on OBS/IBS at build-time unless I overlooked where you rely on something that OBS/IBS can not provide, e.g. screen based testing, remote host access or multi-machine testing. Can you comment why this test should be in openQA tests?

perlpunk commented:
@okurz I have no idea.
I talked to several people about how I should proceed with my first openQA test.
How can I move my test to OBS?

okurz commented:
As discussed on chat:
https://build.opensuse.org/package/show/openSUSE:Factory/sqlite3 for a start with afterwards potential backports into the branched package for all our released products and such. In https://build.opensuse.org/package/view_file/openSUSE:Factory/sqlite3/sqlite3.spec?expand=1 in line 145 you can see "make test" being called so I wonder what the upstream tests already cover or what blind spots your test could cover. In the build log, e.g. https://build.opensuse.org/package/live_build_log/openSUSE:Factory/sqlite3/standard/x86_64 , you can see the output of test results.

I leave the decision to you if you want this PR to have merged anyway or rethink.

okurz commented:
please also provide a verification run for at least openSUSE Tumbleweed.

perlpunk commented:
I squashed, rebased and force-pushed again.
Like discussed on chat, the tests won't hurt, and I can try to figure out how to get the build tests from factory also running in SLES

okurz commented:
yes, but please also provide a verification run for at least openSUSE Tumbleweed. You can easily use http://open.qa/docs/#_triggering_tests_based_on_an_any_remote_git_refspec_or_open_github_pull_request and also skip some other test modules on trigger to reduce the testing time.

perlpunk commented:
I had to add another thing: on tumbleweed, `expect` wasn't installed. force-pushed again

okurz commented:
the test on Tumbleweed looks fine but because I have not seen any "serial terminal" test within https://openqa.opensuse.org/tests/985293 so far could you please test the whole scenario with your module included?

perlpunk commented:
@okurz http://10.161.229.247/tests/105 looks ok (1 failure)
